### PR TITLE
Add column_device_view to orc writer

### LINKED ISSUE #7347: [FEA] Refactor stats_column_desc in cuio
The stats_column_desc structure contains multiple members such as `valid_map_base`, `column_offset`, `column_data_base` etc. that can replaced by `column_device_view`.

Doing so will require changing other functions accordingly such as `GatherColumnStatistics`, `MergeColumnStatistics` used for stats calculation.

### PR DESCRIPTION
**Pull request by kaatish**
This PR adds column_device_view members to EncChunk, DictionaryChunk and StripeDictionary structures which are used in the ORC writer. The idea is to replace members in these structures which replicate the same information. Usage of nvstrdesc_s has also been eliminated in the ORC writer.

Fixes #7347, Addresses #5682, Addresses #7334

### CODE REVIEW

File: cpp/src/io/orc/writer_impl.cu [hunk 1]
```diff
@@ -1074,6 +1027,9 @@ void writer::impl::write(table_view const &table)
       "be specified");
   }
 
+  auto device_columns    = table_device_view::create(table);
```
Comment (kaatish) [inline cpp/src/io/orc/writer_impl.cu:1030]: Will add stream to the create call.

**Comment by vuule** [comment]
> Fixes #7347, #5682 Addresses #7334

You cannot apply "fixes" to multiple issues like this. Has to be "Fixes #7347, fixes #5682, addresses #7334"

**Comment by vuule** [comment]
Should this PR close #5682? I see uses of nvstrdesc_s that are not removed in this PR. I would expect for all nvstrdesc_s declaration to be removed before we close #5682.

File: cpp/src/io/orc/dict_enc.cu [hunk 2]
```diff
@@ -46,14 +47,16 @@ struct dictinit_state_s {
 };
 
 /**
- * @brief Return a 12-bit hash from a byte sequence
+ * @brief Return a 12-bit hash from a string
  */
-static inline __device__ uint32_t nvstr_init_hash(char const *ptr, uint32_t len)
+static inline __device__ uint32_t hash_string(const string_view &val)
```
Comment (vuule) [inline cpp/src/io/orc/dict_enc.cu:52]: again, I think this should be passed by value, but libcudf is not very consistent. Will check.

File: cpp/src/io/orc/dict_enc.cu [hunk 3]
```diff
@@ -233,18 +253,17 @@ __global__ void __launch_bounds__(block_size, 2)
...
     uint32_t ck_row = 0, ck_row_ref = 0, is_dupe = 0;
     if (i + t < nnz) {
-      const char *str1, *str2;
-      uint32_t len1, len2, hash;
-      ck_row     = s->dict[i + t];
-      str1       = ck_data[ck_row].ptr;
-      len1       = static_cast<uint32_t>(ck_data[ck_row].count);
-      hash       = nvstr_init_hash(str1, len1);
-      ck_row_ref = s->dict[(hash > 0) ? s->map.u16[hash - 1] : 0];
+      uint32_t string_length, hash;
```
Comment (vuule) [inline cpp/src/io/orc/dict_enc.cu:256]: please declare at first use (as const, if possible)

**Review by vuule — CHANGES_REQUESTED** [review]
Big fan of the changes. Just needs more span and to add missing docs. And the span use can be pushed to another PR IMO.

Have you run the benchmarks?

**Comment by ttnghia** [comment]
> Fixes #7347, #5682 Addresses #7334

Please add more description. It's not obvious to know what this PR is doing unless we have to click/hover on those links to open these issues.

**Comment by davidwendt** [comment]
> > Fixes #7347, #5682 Addresses #7334
> 
> Please add more description. It's not obvious to know what this PR is doing unless we have to click/hover on those links to open these issues.

Also, the description gets added to the changelog entry so even a simple description here would be helpful.

**Comment by kaatish** [comment]
> Should this PR close #5682? I see uses of nvstrdesc_s that are not removed in this PR. I would expect for all nvstrdesc_s declaration to be removed before we close #5682.

I have removed the link to that issue for now but it talks about removing nvstrdesc_s in the writers only which has been done in this PR.

**Comment by vuule** [comment]
> > Should this PR close #5682? I see uses of nvstrdesc_s that are not removed in this PR. I would expect for all nvstrdesc_s declaration to be removed before we close #5682.
> 
> I have removed the link to that issue for now but it talks about removing nvstrdesc_s in the writers only which has been done in this PR.

@devavret to clarify the issue. My interpretation is that all nvstrdesc_s uses should be replaced.

**Comment by devavret** [comment]
> > > Should this PR close #5682? I see uses of nvstrdesc_s that are not removed in this PR. I would expect for all nvstrdesc_s declaration to be removed before we close #5682.
> > 
> > 
> > I have removed the link to that issue for now but it talks about removing nvstrdesc_s in the writers only which has been done in this PR.
> 
> @devavret to clarify the issue. My interpretation is that all nvstrdesc_s uses should be replaced.

That is true. It just might need one more PR. So it could be an "addresses" here.

**Review by vuule — APPROVED** [review]

File: cpp/src/io/orc/writer_impl.cu [hunk 4]
```diff
@@ -233,9 +184,9 @@ class orc_column_view {
   size_t data_count() const noexcept { return _data_count; }
   size_t null_count() const noexcept { return _null_count; }
   bool nullable() const noexcept { return (_nulls != nullptr); }
-  void const *data() const noexcept { return _data; }
+  // void const *data() const noexcept { return _data; }
```
Comment (devavret) [inline cpp/src/io/orc/writer_impl.cu:187]: Cleanups?

File: cpp/src/io/orc/orc_gpu.h [hunk 5]
```diff
@@ -313,6 +313,10 @@ void EncodeStripeDictionaries(StripeDictionary *stripes,
                               detail::device_2dspan<encoder_chunk_streams> enc_streams,
                               rmm::cuda_stream_view stream = rmm::cuda_stream_default);
 
+void set_chunk_columns(const table_device_view &view,
```
Comment (devavret) [inline cpp/src/io/orc/orc_gpu.h:316]: doc

File: cpp/src/io/orc/writer_impl.cu (hunk 4 above)
Comment (kaatish) [reply cpp/src/io/orc/writer_impl.cu:187]: Yes, missed this one.

**Review by devavret — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**